test/signer.ts

```diff
@@ -0,0 +1,504 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import * as assert from 'assert';
+import * as crypto from 'crypto';
+import * as sinon from 'sinon';
+import * as url from 'url';
+
+import {
+  UrlSigner,
+  SigningError,
+  AuthClient,
+  BucketI,
+  FileI,
+  SignerGetSignedUrlConfig,
+  PATH_STYLED_HOST,
+} from '../src/signer';
+
+describe('signer', () => {
+  const BUCKET_NAME = 'bucket-name';
+  const FILE_NAME = 'file-name.png';
+  const CLIENT_EMAIL = 'client-email';
+
+  let sandbox: sinon.SinonSandbox;
+  beforeEach(() => (sandbox = sinon.createSandbox()));
+  afterEach(() => sandbox.restore());
+
+  describe('UrlSigner', () => {
+    let authClient: AuthClient;
+    let bucket: BucketI;
+    let file: FileI;
+
+    const NOW = new Date('2019-03-18T00:00:00Z');
+    let fakeTimers: sinon.SinonFakeTimers;
+
+    beforeEach(() => (fakeTimers = sinon.useFakeTimers(NOW)));
+    afterEach(() => fakeTimers.restore());
+
+    beforeEach(() => {
+      authClient = {
+        sign: async (_blobToSign: string) => 'signature',
+        getCredentials: async () => ({client_email: CLIENT_EMAIL}),
+      };
+      bucket = {name: BUCKET_NAME};
+      file = {name: FILE_NAME};
+    });
+
+    it('should construct a UrlSigner', () => {
```

Comment:
Just for style-consistency, we would normally wrap constructor tests in a describe:

```js
describe('UrlSigner', () => {
  it('should localize authClient', () => {});

  it('should localize bucket', () => {});

  it('should localize file', () => {});
});
```

Seems excessive when you compare it to just the single test, but multiple tests should highlight to other maintainers that if they change something or add a new property, it needs another test.

Reply:
done!